src/Microsoft.Tye.Hosting/BuildWatcher.cs

```diff
@@ -0,0 +1,293 @@
+
+                    var solutionBatch = new Dictionary<string, List<BuildRequest>>(); //  store the list of promises
+                    var projectBatch = new Dictionary<string, List<BuildRequest>>();
+
+                    foreach (var request in requests)
+                    {
+                        if (solution?.ProjectShouldBuild(request.ProjectFilePath) == true)
+                        {
+                            if (!solutionBatch.ContainsKey(request.ProjectFilePath))
+                            {
+                                solutionBatch.Add(request.ProjectFilePath, new List<BuildRequest>());
+                            }
+
+                            solutionBatch[request.ProjectFilePath].Add(request);
+                        }
+                        else
+                        {
+                            // this will also prevent us building multiple times if a project is used by multiple services
+                            if (!projectBatch.ContainsKey(request.ProjectFilePath))
+                            {
+                                projectBatch.Add(request.ProjectFilePath, new List<BuildRequest>());
+                            }
+
+                            projectBatch[request.ProjectFilePath].Add(request);
+                        }
+                    }
+
+                    async Task WithRequestCompletion(IEnumerable<BuildRequest> requests, Func<Task<int>> buildFunc)
+                    {
+                        try
+                        {
+                            int exitCode = await buildFunc();
+
+                            foreach (var request in requests)
+                            {
+                                request.Complete(exitCode);
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            foreach (var request in requests)
+                            {
+                                request.Complete(ex);
+                            }
+                        }
+                    }
+
+                    var tasks = new List<Task>();
+
+                    if (solutionBatch.Any())
+                    {
+                        var targets = String.Join(",", solutionBatch.Keys.Select(key => GetProjectName(solution!, key)));
+
+                        tasks.Add(
+                            WithRequestCompletion(
+                                solutionBatch.Values.SelectMany(x => x),
+                                async () =>
+                                {
+                                    logger.LogInformation("Build Watcher: Building {Targets} of solution {SolutionPath}...", targets, solutionPath);
+
+                                    var buildResult = await ProcessUtil.RunAsync("dotnet", $"msbuild {solutionPath} -target:{targets}", throwOnError: false, workingDirectory: workingDirectory, cancellationToken: cancellationToken);
```

Comment:
Note for #1438, this code is not replacing "special characters" with an `_` in the targets.

![image](https://user-images.githubusercontent.com/7330891/185206890-8659e87f-688e-443e-8ac0-6819911c8f3f.png)
https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-build-specific-targets-in-solutions-by-using-msbuild-exe?view=vs-2022